src/gg.regression.unity.bots/Runtime/Scripts/GenericBots/RGMonkeyBot.cs

```diff
@@ -53,55 +61,133 @@ void Update()
+                _mouseBtnsBuf.Clear();
+                foreach (var inp in _actionsBuf.Where(act => act.Performed).SelectMany(act => act.Inputs))
+                {
+                    if (inp is MousePositionInput)
+                    {
+                        haveMousePos = true;
+                    } else if (inp is MouseButtonInput mbInput)
+                    {
+                        if (mbInput.IsPressed)
+                        {
+                            _mouseBtnsBuf.Add(mbInput.MouseButton);
+                        }
+                    } else if (inp is LegacyKeyInput keyInput)
+                    {
+                        if (keyInput.KeyCode >= KeyCode.Mouse0 && keyInput.KeyCode <= KeyCode.Mouse6)
+                        {
+                            _mouseBtnsBuf.Add(keyInput.KeyCode - KeyCode.Mouse0);
+                        }
+                    }
+                }
+
+                if (haveMousePos && _mouseBtnsBuf.Count > 0)
                 {
-                    _validActionGroups.Add(action.ActionGroup);
-                    var instBuf = _actionsByGroup[action.ActionGroup][action];
-                    foreach (var inst in instances)
+                    if (Random.Range(0, 2) == 1) // 50% chance of attempting a click event
                     {
-                        instBuf.Add(inst);
+                        foreach (var btn in _mouseBtnsBuf)
+                        {
+                            new MouseButtonInput(btn, false).Perform();
+                        }
+                        didHeuristic = true;
                     }
                 }
             }
-            foreach (int grp in _validActionGroups)
+            if (didHeuristic)
             {
-                _validActionGroupsList.Add(grp);
+                ResetState();
+                return;
             }
             
-            // 1. Randomly choose an action group
-            // 2. Choose one action instance to perform from each action in the group
-            if (_validActionGroupsList.Count > 0)
+            ResetState();
+
+            // Compute the set of valid actions
+            foreach (var actionInstance in RGActionManager.GetValidActions())
+            {
+                var entry = new MonkeyBotActionEntry()
+                {
+                    ActionInstance = actionInstance,
+                    Parameter = actionInstance.BaseAction.ParameterRange.RandomSample(),
+                    Performed = false
+                };
+                entry.Inputs = new List<RGActionInput>(actionInstance.GetInputs(entry.Parameter));
+                if (entry.Inputs.Count > 0)
+                {
+                    _actionsBuf.Add(entry);
+                }
+            }
+
+            // Randomly perform actions from the list 
+            // This repeatedly selects actions whose inputs do not overlap
+            // with the inputs that have already been performed.
+            for (;;)
```

Comment:
while(true) loops scare me

let's make this something like

`do {....} while(_remainingActionsBuf.Count >0)`

Reply:
@svolokh other changes look good as far as I can tell

just this last comment remaining from my end

Reply:
Updated this one to be do-while as suggested.